chapter_computational-performance/auto-parallelism.md

```diff
@@ -1,92 +1,183 @@
-run(x_gpu)
+run(x_gpu1)  # Warm-up both devices
+run(x_gpu2)
 npx.waitall()  
 
-with d2l.Benchmark('CPU time'):
-    run(x_cpu)
+with d2l.Benchmark('GPU1 time'):
+    run(x_gpu1)
     npx.waitall()
 
-with d2l.Benchmark('GPU time'):
-    run(x_gpu)
+with d2l.Benchmark('GPU2 time'):
+    run(x_gpu2)
     npx.waitall()
 ```
 
+```{.python .input}
+#@tab pytorch
+run(x_gpu1)
+run(x_gpu2)  # Warm-up all devices
+torch.cuda.synchronize(devices[0])
+torch.cuda.synchronize(devices[1])
+
+with d2l.Benchmark('GPU 1 time'):
+    run(x_gpu1)
+    torch.cuda.synchronize(devices[0])
+
+with d2l.Benchmark('GPU 2 time'):
+    run(x_gpu2)
+    torch.cuda.synchronize(devices[1])
+```
+
+:begin_tab:`mxnet`
 If we remove the `waitall()` between both tasks the system is free to parallelize computation on both devices automatically.
+:end_tab:
+
+:begin_tab:`pytorch`
+If we remove the `torch.cuda.synchronize()` between both tasks the system is free to parallelize computation on both devices automatically.
+:end_tab:
 
 ```{.python .input}
-with d2l.Benchmark('CPU & GPU'):
-    run(x_cpu)
-    run(x_gpu)
+with d2l.Benchmark('GPU1 & GPU2'):
+    run(x_gpu1)
+    run(x_gpu2)
     npx.waitall()
 ```
 
-In the above case the total execution time is less than the sum of its parts, since MXNet automatically schedules computation on both CPU and GPU devices without the need for sophisticated code on behalf of the user. 
+```{.python .input}
+#@tab pytorch
+with d2l.Benchmark('GPU1 & GPU2'):
+    run(x_gpu1)
+    run(x_gpu2)
+    torch.cuda.synchronize()
+```
+
+:begin_tab:`mxnet`
+In the above case the total execution time is less than the sum of its parts, since MXNet automatically schedules computation on both GPU devices without the need for sophisticated code on behalf of the user.
+:end_tab:
+
+:begin_tab:`pytorch`
+In the above case the total execution time is less than the sum of its parts, since PyTorch automatically schedules computation on both GPU devices without the need for sophisticated code on behalf of the user.
+:end_tab:
 
 ## Parallel Computation and Communication
-
 In many cases we need to move data between different devices, say between CPU and GPU, or between different GPUs. This occurs e.g., when we want to perform distributed optimization where we need to aggregate the gradients over multiple accelerator cards. Let us simulate this by computing on the GPU and then copying the results back to the CPU.
 
 ```{.python .input}
 def copy_to_cpu(x):
     return [y.copyto(npx.cpu()) for y in x]
 
-with d2l.Benchmark('Run on GPU'):
-    y = run(x_gpu)
+with d2l.Benchmark('Run on GPU1'):
+    y = run(x_gpu1)
     npx.waitall()
 
 with d2l.Benchmark('Copy to CPU'):
     y_cpu = copy_to_cpu(y)
     npx.waitall()
 ```
 
+```{.python .input}
+#@tab pytorch
+def copy_to_cpu(x, non_blocking=False):
+    return [y.to('cpu', non_blocking=non_blocking) for y in x]
+
+with d2l.Benchmark('Run on GPU1'):
+    y = run(x_gpu1)
+    torch.cuda.synchronize()
```

Comment:
what's the behavior if we do not pass any argument here

Reply:
It will use the current device. See [here](https://pytorch.org/docs/stable/cuda.html#torch.cuda.synchronize).

Reply:
Thanks. Could you add such explanations in the text? This will be useful

Reply:
Sure! Will do that.

Reply:
Added in #1602